### PR TITLE
`Board` docs

### PR DESCRIPTION
**Pull request by mattjperez**
Changes to crate documentation for developer use. 
Method descriptions try to mirror the [Python SDK](https://python.viam.dev/autoapi/viam/components/index.html).

This PR highlights different options for how documentation can be added and rendered, as well as the directory structure changes (individual `mod.rs` or using the root `lib.rs`).

If this is approved, I'll add docs for the other components.

use `make doc-open` to generate docs and view in browser

### CODE REVIEW

File: src/common/board.rs
Comment (gvaradarajan) [inline src/common/board.rs:75]: Should specify here (and in L. 73) that the domain for duty cycle is a percent value float between 0 and 1

File: src/common/board.rs
Comment (gvaradarajan) [inline src/common/board.rs:80]: This is not true, setting the frequency to 0 unregisters the pin and PWM channel from the timer and removes the PWM signal

File: src/common/mod.rs
Comment (gvaradarajan) [inline src/common/mod.rs:1]: Some of these are actually natively-included general purpose drivers (gpio_motor, adxl, mpu), so maybe we should note those separately?

File: src/lib.rs
Comment (gvaradarajan) [inline src/lib.rs:1]: I think this reorganization is definitely more intuitive at the very least, but it might be two big of a change as we're going into beta? We might want to hold off on this

**Review by gvaradarajan — CHANGES_REQUESTED** [review]
Thought I'd put in a review anyway, but it's really up to @npmenard whether this goes in. I like the PR for the most part.

File: src/common/board.rs
Comment (mattjperez) [reply src/common/board.rs:80]: Definitely something that should be included. Thanks for clarifying.

File: src/lib.rs
Comment (mattjperez) [reply src/lib.rs:1]: Good to know it feels more intuitive. 

While it definitely feels like a big change, it doesn't affect any of the main logic or imports.

There is an alternative though if you'd still like the docs without altering the structure.

https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#the-doc-attribute

This would just be missing the compartmentalizing advantages of having inline documentation.

File: src/common/board.rs
Comment (acmorrow) [inline src/common/board.rs:52]: Very minor, but could these be offset by a blank line so the code and comments don't run together? Or will the rust formatter just shove them backtogether? Personally, I find it much more readable as:

```
    /// Set a pin to high or low
    fn set_gpio_pin_level(&mut self, pin: i32, is_high: bool) -> anyhow::Result<()>;

    /// Return the current [BoardStatus] of the board
    fn get_board_status(&self) -> anyhow::Result<common::v1::BoardStatus>;
```

It seems, at least, to be the style within the rust std lib: https://doc.rust-lang.org/src/core/alloc/mod.rs.html#164

Comment (mattjperez) [reply src/common/board.rs:52]: Sure!

We may want to look at adding a `rustfmt.toml` file in the future to enforce these things. 

Though it seems this specifically is still unstable. [Link](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#blank_lines_lower_bound)

File: src/common/board.rs
Comment (randhid) [inline src/common/board.rs:90]: we... maybe be getting rid of i2c altogether and having the sensors handle them. so I'd hold off on that  pr to go in before doing too much there - the i2c call is not part of our api.

**Review by randhid — CHANGES_REQUESTED** [review]

File: src/common/board.rs
Comment (mattjperez) [reply src/common/board.rs:90]: Ok. I don't think I touched anything specifically with i2c though, this change was just me moving `FakeBoard`'s struct definition below the `Board` trait definition, to group related things closer together.

Comment (mattjperez) [reply src/common/board.rs:90]: Ahh I think I see what you're saying. The `Board` trait itself defines some functions for i2c that aren't part of the Board API. Yeah I was only putting docstrings on methods that I found equivalents for in the Python SDK, which I believe does accurately reflect the api.

File: src/common/board.rs
Comment (mattjperez) [reply src/common/board.rs:80]: I looked at `esp32::board`'s implementation of this, and it doesn't unregister the PWM channel from the timer. It just sets the pin's `pwm_driver` to `None`. 

There is a `Drop` implementation for `PwmDriver` but I don't think it's ever actually dropped since `pin.pwm_driver` is `Option<PwmDriver<'static>>`. If so, we need to change how `0` is handled. Though this would be outside the scope of this PR.

File: src/common/board.rs
Comment (randhid) [reply src/common/board.rs:100]: sort of - here's the ground truth for the [board API](https://github.com/viamrobotics/api/blob/main/proto/viam/component/board/v1/board.proto), python has a lot of client side-only helper functions, which might be going away.

File: src/common/board.rs
Comment (gvaradarajan) [reply src/common/board.rs]: I've confirmed via print statement that it is in fact dropped

File: src/common/board.rs
Comment (acmorrow) [reply src/common/board.rs:100]: Since it is just code motion though I don't think this PR should be blocked on any i2c work. Is that ok with you @randhid ?

File: src/common/board.rs
Comment (acmorrow) [inline src/common/board.rs:71]: Looks like no doc comment for this function. Per the other discussion about i2c I recognize there may be upcoming changes, but it feels OK to me to document what exists here now, even if it ends up moving elsewhere later.

Comment (mattjperez) [reply src/common/board.rs:71]: I agree

File: src/common/board.rs
Comment (randhid) [reply src/common/board.rs:82]: totally okay.

**Review by randhid — APPROVED** [review]
LGTM

File: src/common/board.rs
Comment (npmenard) [inline src/common/board.rs:7]: nit: would just import proto::component so we know PowerMode is a generated type

File: src/common/board.rs
Comment (npmenard) [inline src/common/board.rs:37]: common::v1::BoardStatus

File: src/common/board.rs
Comment (npmenard) [inline src/common/board.rs:64]: Can you add the condition when this function can fail?

File: src/common/board.rs
Comment (npmenard) [inline src/common/board.rs:67]: same

File: src/common/board.rs
Comment (npmenard) [inline src/common/board.rs:79]: nit: Emulate functionality of a Board for test purpose. 
I think we should maybe hide this from docs

**Review by npmenard — CHANGES_REQUESTED** [review]
done with first pass

**Review by gvaradarajan — APPROVED** [review]
Needs a minor adjustment due to the two most recently merged PRs, so make sure to rebase in main and make the requisite changes. Otherwise LGTM

File: src/common/board.rs
Comment (mattjperez) [reply src/common/board.rs:79]: To clarify, hide the description or hide the `FakeBoard` itself?
For the first we can just put a regular comment, for the second we can use `#[doc(hidden)]`

File: src/common/board.rs
Comment (mattjperez) [reply src/common/board.rs:64]: In our `FakeBoard`'s impl or `Board`, it never fails and just returns `0.0` if the pin isn't present in the `pwm_pins` map. Should `Board::get_pwm_duty`'s description say it will return `0.0` or say that it's implementation specific?

File: src/common/board.rs
Comment (npmenard) [reply src/common/board.rs:79]: let's just hide it altogether

File: src/common/mod.rs
Comment (gvaradarajan) [inline src/common/mod.rs:50]: needs `pub mod servo;` and `pub mod gpio_servo;`

**Review by gvaradarajan — CHANGES_REQUESTED** [review]

File: src/common/mod.rs
Comment (mattjperez) [reply src/common/mod.rs:52]: had it in a local commit. pushed it

**Review by gvaradarajan — APPROVED** [review]
Once you rebase again, you will also need `pub mod power_sensor;` in common/mod.rs

**Review by npmenard — APPROVED** [review]
LGTM

**Review by npmenard — APPROVED** [review]
go for merge

**Comment by mattjperez** [comment]
@viamrobotics/micro-rdk-team could I get help with the merge? Contributors don't have write access.

**Pull request merged**
